README.md
# ocaml-SeqBox
Implementation of [SeqBox](https://github.com/MarcoPon/SeqBox) in OCaml

The original pharsing was "Port of SeqBox to OCaml", but since no direct porting/translation was actually done due to the differences between Python 3(implementation language of SeqBox) and OCaml, which also force the architecture and design of this software(ocaml-SeqBox) to be independently developed, thus this is only an implementation(not a port or translation) of SeqBox according to its technical specifications.
This is mainly to address the different licenses being used(SeqBox was using AGPL 3.0 at the time of writing while this project uses 3-Clause BSD license).

Official SeqBox Repo - https://github.com/MarcoPon/SeqBox

Table of Contents
=================

   * [ocaml-SeqBox](#ocaml-seqbox)
      * [Acknowledgement](#acknowledgement)
      * [Getting started](#getting-started)
         * [Recordings](#recordings)
         * [Installation](#installation)
      * [Notes](#notes)
      * [Tips](#tips)
      * [Gotchas](#gotchas)
      * [Helpers](#helpers)
      * [Links](#links)
      * [Possibly useful additional features of ocaml-SeqBox(possibly not yet in official SeqBox)](#possibly-useful-additional-features-of-ocaml-seqboxpossibly-not-yet-in-official-seqbox)
      * [Technical Specification](#technical-specification)
         * [Common blocks header:](#common-blocks-header)
         * [Block 0](#block-0)
         * [Blocks &gt; 0 &amp; &lt; last:](#blocks--0---last)
         * [Blocks == last:](#blocks--last)
         * [Versions:](#versions)
         * [Metadata encoding:](#metadata-encoding)
            * [IDs](#ids)
            * [Features currently NOT planned to be implemented](#features-currently-not-planned-to-be-implemented)
      * [License](#license)

Created by [gh-md-toc](https://github.com/ekalinin/github-markdown-toc)

## Acknowledgement
I would like to thank [Marco](https://github.com/MarcoPon/) (author of the official SeqBox) for discussing and clarifying several aspects of his project, and providing me with test data.

I would like to thank [Ming](https://github.com/mdchia/) for his feedback on the documentation, UX design, and several other general aspects of this project. And also his help on testing the building and installation of osbx on macOS.

## Getting started
### Recordings
You can view the recordings [here](https://github.com/darrenldl/ocaml-SeqBox/wiki/Recordings)

### Installation
Osbx 1.2.4 is available through OPAM.
```
opam install osbx
```
There are four modes/commands for osbx currently : encode, decode, rescue, show

You can consult the man pages of osbx itself and all the four commands via
```
osbx        --help
osbx encode --help
osbx decode --help
osbx rescue --help
osbx show   --help
```

## Notes
Version 1.2.4 is considered to be feature complete and mature enough for production use, but any bug reports or suggestions are very welcome - just open an issue!

Contributions are welcome as well, but note that by submitting a contribution, you agree that your code will be licensed under the 3-Clause BSD license.

No major active development will occur, but since 1.2.4 was designed to be quite scriptable via the options, you can always write helper scripts for advanced features. You can check out the ones I am writing [here](https://github.com/darrenldl/osbx-helpers).

CRC-CCITT is currently implemented in pure OCaml and is translated from the implementation in [libcrc](https://github.com/lammertb/libcrc)
  - See src/crcccitt.ml, src/crcccitt.mli for the OCaml implementation
    - The translated source code is under the same MIT license used by and stated in libcrc source code
  - See libcrc\_crcccitt/crcccitt.c, libcrc\_crcccitt/checksum.h for the source code from libcrc used for the translation

Exact behaviours in non-standard cases are not specified in official SeqBox technical specification
  - See [specification](SPECS.md) of ocaml-SeqBox for details on how ocaml-SeqBox behaves(if you care about undefined behaviour those sort of things)

Hashing libraries
  - Nocrypto for SHA1, SHA256, SHA512
  - Digestif for BLAKE2B\_512

## Tips
See [wiki page](https://github.com/darrenldl/ocaml-SeqBox/wiki/Tips)

## Gotchas
See [wiki page](https://github.com/darrenldl/ocaml-SeqBox/wiki/Gotchas)

## Helpers
See [helpers](https://github.com/darrenldl/osbx-helpers)

## Links
[Wiki](https://github.com/darrenldl/ocaml-SeqBox/wiki)

[Index of source code](INDEX.md)

[Changelog](CHANGELOG.md)

[Todo/wishlist](TODO.md)

## Possibly useful additional features of ocaml-SeqBox(possibly not yet in official SeqBox)
  - Allows random ordering in sbx container
    - This also means block corruption will not stop the decoding process
  - Allows duplicate metadata/data blocks to exist within one sbx container
    - This means you can concatenate multiple copies of sbx container together directly to increase chance of recovery in case of corruption

## Technical Specification
The following specification is copied directly from the official specification (with possible slight modifications).

Also see section "Features currently NOT planned to be implemented" for features ocaml-SeqBox is probably not going to have.

Byte order: Big Endian
### Common blocks header:

| pos | to pos | size | desc                                |
|---- | ---    | ---- | ----------------------------------- |
|  0  |      2 |   3  | Recoverable Block signature = 'SBx' |
|  3  |      3 |   1  | Version byte |
|  4  |      5 |   2  | CRC-16-CCITT of the rest of the block (Version is used as starting value) |
|  6  |     11 |   6  | file UID                            |
| 12  |     15 |   4  | Block sequence number               |

### Block 0

| pos | to pos   | size | desc             |
|---- | -------- | ---- | ---------------- |
| 16  | n        | var  | encoded metadata |
|  n+1| blockend | var  | padding (0x1a)   |

### Blocks > 0 & < last:

| pos | to pos   | size | desc             |
|---- | -------- | ---- | ---------------- |
| 16  | blockend | var  | data             |

### Blocks == last:

| pos | to pos   | size | desc             |
|---- | -------- | ---- | ---------------- |
| 16  | n        | var  | data             |
| n+1 | blockend | var  | padding (0x1a)   |

### Versions:
N.B. Current versions differs only by blocksize.

| ver | blocksize | note    |
|---- | --------- | ------- |
|  1  | 512       | default |
|  2  | 128       |         |
|  3  | 4096      |         |

### Metadata encoding:

| Bytes | Field | 
| ----- | ----- |
|    3  | ID    |
|    1  | Len   |
|    n  | Data  |

#### IDs

| ID | Desc |
| --- | --- |
| FNM | filename (utf-8) |
| SNM | sbx filename (utf-8) |
| FSZ | filesize (8 bytes) |
| FDT | date & time (8 bytes, seconds since epoch) |
| SDT | sbx date & time (8 bytes) |
| HSH | crypto hash (using [Multihash](http://multiformats.io) protocol) |
| PID | parent UID (*not used at the moment*)|

Supported crypto hashes since 1.1.0 are
  - SHA1
  - SHA256
  - SHA512
  - BLAKE2B\_512

#### Features currently NOT planned to be implemented
  - Data hiding (XOR encoding/decoding in official seqbox)
    - Provides neither sufficiently strong encryption nor sufficient stealth for any serious attempt to hide/secure data
    - You should use the appropriate tools for encryption

## License
The following files directly from [libcrc](https://github.com/lammertb/libcrc) are under the MIT License(see license text in each of the code files)
  - libcrc\_crcccitt/crcccitt.c
  - libcrc\_crcccitt/checksum.h

The following files translated from libcrc source code are under the same MIT License as used by libcrc and as stated in libcrc source code, the license text of the crcccitt.c is copied over to src/crcccitt.ml as well
  - src/crcccitt.ml
  - src/crcccitt.mli

The files in tests folder copied from official SeqBox are under its license, which is MIT as of time of writing
  - tests/SeqBox/*

All remaining files are distributed under the 3-Clause BSD license as stated in the LICENSE file


CHANGELOG.md
## 1.2.4  (newest version on OPAM)
  - Adapted to Angstrom 0.7.0 API changes
    - This is completely internal
  - Added --only-pick-uid option to rescue mode
    - This allows only picking sbx blocks of specific uid when rescuing
    - This is useful when you only only want to rescue one specific file
  - Renamed --only-pick option to --only-pick-block in rescue mode
  
## 1.2.3  (published on OPAM)
  - Message text polishing
  - Fixed division by zero error/exception in progress reporting when number of total units is 0
    - The error/exception affects all four modes when dealing with files of size 0 bytes
  - Added only-pick option to rescue mode
    - This allows asking osbx to only pick some type(s) of blocks during rescuing
  - Added seek position checking code for rescue log file
    - This prevents showing "Unknown error" in rescue mode when logged bytes processed field exceeds the file size
  - File UID hex is printed in upper case
  - Added from, to options to rescue mode
    - This allows specifying range of file to scan for data rescuing
  - Added from, to options to show mode(previously only skip-to option exists for show)
    - This allows specifying range of file to scan for metadata
  - Added force-misalign option to rescue and show mode
    - This allows disabling automatic rounding down of from byte parameter

## 1.2.2  (published on OPAM)
  - Optimized decoding code to speed up reference block scanning
    - Avoids decoding the sbx block in full when already has a block of that type
  - Added no-meta option to decode mode
    - If enabled, saves time by not looking for metadata block actively
  - Added code to handle negative parameters
    - decode
      - show-fail-max
    - show
      - find-max
      - skip-to

## 1.2.1  (published on OPAM)
  - Optmized CRC-CCITT
    - Changed fake\_uint16 to use int instead of int64
    - This makes encoding and decoding significantly faster
    - Other block related operations are also faster in general
  - Added silent option to all modes
    - Only affects progress text printing

## 1.2.0  (superseded by 1.2.1, not published on OPAM)
  - Fixed slightly off progress counting in Decode
  - Optimized decoding code to avoid searching for a reference block twice
  - Added command line option for decode and show
    - Added --show-fail-max for decode
    - Added --find-max      for show
  - Updated show command to also show sbx container version
  - Added position tracking to show mode
  - Added --skip-to for show for skipping to some byte
  - Improved progress reporting
    - Replaced fraction with progress bar
    - Made the rate more readable
    - General overhaul to improve quality
  - Optimized performance of progress printing code and log writing code

## 1.1.1  (published on OPAM)
  - Removed dependency on Core\_kernel
    - 1.1.1 uses a custom replacement of Core\_kernel.protect and OCaml Pervasives channels

## 1.1.0  (superseded by 1.1.1, not published on OPAM)
  - Added "show" command
  - Switched from using Core to Core\_kernel
    - Now osbx can be compiled on Windows
  - Switched to a pure OCaml implementation of CRC-CCITT
    - Implementation is translated from libcrc implementation
  - Added support for multiple hash functions
    - added support for sha1 sha256 sha512 and blake2b-512
    - in prior versions, only sha256 was supported
  - Encoding, decoding now try to be "smart" (see their man pages)
  - Fixed Multihash encoding, decoding of BLAKE2 hashes

## 1.0.1  (published on OPAM)
  - Changed rescue mode to use the original bytes of the file, rather than regenerating the block bytes itself
  - Fixed jbuild file to work on macOS when building
  - Added graceful handling of Ctrl-C breaks in rescue mode
    - log file stays valid after being interrupted
    - in 1.0.0, log file may be empty/broken when osbx is interrupted during rescuing
  - Better progress reporting for encoding, decoding and rescuing
  - Fixed progress reporting issues with rescuing
    - in 1.0.0, rescue mode only does a progress report when it outputs a block, rather than after scanning the block bytes
    - this makes osbx outputs no log when the scanned file section is large and contains no valid blocks
  - Fixed log writing with rescuing
    - in 1.0.0, rescue mode only writes to log file when a valid block is found
    - now it is changed to update log on bytes read, not on blocks written
  - Updated file size retrieval function to act correctly when dealing with block devices
    - Changed File\_utils.getsize to use Core.In\_channel.length instead of Unix.LargeFile.stat
  - Added progress reporting for hash in decode mode

## 1.0.0  (published on OPAM)
  - Base version


src/osbx.ml
open Cmdliner

let help_secs = [ `S Manpage.s_common_options
                ; `P "These options are common to all comamnds."
                ; `S "MORE HELP"
                ; `P "Use `$(mname) $(i,COMMAND) --help' for help on a single command."
                ; `S Manpage.s_authors
                ; `P "Darren Ldl <darrenldldev@gmail.com>"
                ; `S Manpage.s_bugs
                ; `P "Report bugs at ocaml-SeqBox GitHub page via issues (https://github.com/darrenldl/ocaml-SeqBox)"
                ]
;;

let sbx_version =
  let doc = "Sbx container version, one of : 1(bs=512 bytes) 2(bs=128 bytes) 3(bs=4096 bytes) where bs=sbx block size" in
  let open Sbx_specs in
  Arg.(value
       & opt
         (enum [("1", (`V1:version));
                ("2", (`V2:version));
                ("3", (`V3:version))])
         (`V1:version)
       & info ["sbx-version"]
         ~docv:"SBX-VERSION"
         ~doc)
;;

let silent =
  let doc = "One of : 0(show everything), 1(only show progress stats when done) 2(show nothing).
  This only affects progress text printing." in
  let open Progress_report in
  Arg.(value
       & opt
         (enum [("0", L0);
                ("1", L1);
                ("2", L2)])
         L0
       & info ["s"; "silent"]
         ~docv:"LEVEL"
         ~doc)
;;

let force =
  let doc = "Force overwrite even if $(docv) exists" in
  Arg.(value & flag & info ["f"; "force"] ~docv:"OUT" ~doc)
;;

let force_misalign =
  let doc = "Disable automatic rounding down of $(docv).
  This is not normally used and is only intended for data recovery or related purposes." in
  Arg.(value & flag & info ["force-misalign"] ~docv:"FROM-BYTE" ~doc)
;;

let hash =
  let open Multihash in
  let doc =
    "Hash function to use, one of (case-insensitive) : sha1 sha256(default) sha512 blake2b-512" in
  Arg.(value & opt (some string) None & info ["hash"] ~docv:"HASH-TYPE" ~doc)
;;

let from_byte =
  let doc = Printf.sprintf "Start from byte $(docv). The position is automatically rounded down to the closest multiple of %d bytes.
  If not specified, defaults to the start of file.
  Negative values are treated as 0.
  If $(docv) exceeds the largest possible position(file size - 1), then it will be treated as (file size - 1).
  The rounding procedure is applied after all auto-adjustments."
      Param.Common.block_scan_alignment in
  Arg.(value & opt (some int64) None & info ["from"; "skip-to"] ~docv:"FROM-BYTE" ~doc)
;;

let to_byte =
  let doc = "Last position to try to decode a block.
  If not specified, defaults to the end of file.
  Negative values are treated as 0.
  If $(docv) is smaller than FROM-BYTE, then it will be treated as FROM-BYTE." in
  Arg.(value & opt (some int64) None & info ["to"] ~docv:"TO-BYTE" ~doc)
;;

let default_cmd =
  let doc = "a SeqBox implementation written in OCaml" in
  let sdocs = Manpage.s_common_options in
  let exits = Term.default_exits in
  let man = help_secs in
  (Term.(ret (const (`Help (`Pager, None)))),
   Term.info "osbx" ~version:Osbx_version.version ~doc ~sdocs ~exits ~man
  )
;;

let encode_cmd =
  let open Osbx_encode in
  let doc = "Encode file" in
  (Term.(const encode $ silent $ force $ no_meta $ sbx_version $ uid $ hash $ in_file $ out_file),
   Term.info "encode" ~doc
  )
;;

let decode_cmd =
  let open Osbx_decode in
  let doc = "Decode sbx container" in
  (Term.(const decode $ silent $ force $ no_meta $ show_max $ in_file $ out_file),
   Term.info "decode" ~doc
  )
;;

let rescue_cmd =
  let open Osbx_rescue in
  let doc = "Rescue sbx data from file" in
  (Term.(const rescue $ silent $ only_pick_block $ only_pick_uid $ from_byte $ to_byte $ force_misalign $ in_file $ out_dir $ log_file),
   Term.info "rescue" ~doc
  )
;;

let show_cmd =
  let open Osbx_show in
  let doc = "Search for and print metadata in sbx container (or file)" in
  (Term.(const show $ silent $ find_max $ from_byte $ to_byte $ force_misalign $ in_file),
   Term.info "show" ~doc
  )
;;

let () =
  (* catch Ctrl-C breaks *)
  Sys.catch_break true;
  Term.exit @@ Term.eval_choice default_cmd [encode_cmd; decode_cmd; rescue_cmd; show_cmd]
;;
